sled-agent/src/services.rs

```diff
@@ -2688,56 +2685,103 @@ impl ServiceManager {
         Ok(running_zone)
     }
 
-    // Populates `existing_zones` according to the requests in `services`.
-    async fn initialize_omicron_zones_locked(
+    // Ensures that a single Omicron zone is running.
+    //
+    // This method is NOT idempotent.
+    //
+    // - If the zone already exists, in any form, it is fully removed
+    // before being initialized. This is primarily intended to remove "partially
+    // stopped/started" zones with detritus from interfering with a new zone
+    // being launched.
+    // - If zones need time to be synchronized before they are initialized
+    // (e.g., this is a hard requirement for CockroachDb) they can check the
+    // `time_is_synchronized` argument.
+    // - `all_u2_pools` provides a snapshot into durable storage on this sled,
+    // which gives the storage manager an opportunity to validate the zone's
+    // storage configuration against the reality of the current sled.
+    async fn start_omicron_zone(
         &self,
-        existing_zones: &mut BTreeMap<String, RunningZone>,
-        requests: &Vec<OmicronZoneConfigLocal>,
-    ) -> Result<(), Error> {
-        if let Some(name) = requests
-            .iter()
-            .map(|request| request.zone.zone_name())
-            .duplicates()
-            .next()
+        zone: &OmicronZoneConfig,
+        time_is_synchronized: bool,
+        all_u2_pools: &Vec<ZpoolName>,
+    ) -> Result<OmicronZone, Error> {
+        // Ensure the zone has been fully removed before we try to boot it.
+        //
+        // This ensures that old "partially booted/stopped" zones do not
+        // interfere with our installation.
+        self.ensure_removed(&zone).await?;
+
+        // If this zone requires timesync and we aren't ready, fail it early.
+        if zone_requires_timesync(&zone.zone_type) && !time_is_synchronized {
```

Comment:
Tiny nit - `zone_requires_timesync` might be a bit cleaner as a method on `OmicronZoneType`?

Reply:
Agreed, updated to become a method there.